### PR TITLE
fern: inference-time mirror-y TTA on SOTA stack

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The DrivAer family of cars has approximate bilateral (left-right) symmetry about the xz-plane (y=0). At inference/validation time, we can exploit this symmetry via **test-time augmentation (TTA)**: run the model on both the original geometry and the y-mirrored geometry, then average the predictions. This should:

1. **Reduce random variance** in per-batch predictions by averaging two views of each car.
2. **Improve generalization** by cancelling out any learned asymmetric biases in the model.
3. **Directly attack tau_y and tau_z** — lateral wall shear (tau_y) and vertical wall shear (tau_z) are the components with the biggest AB-UPT gap. Mirror-y symmetry constrains: `tau_y(x, y, z) = -tau_y(x, -y, z)` and `tau_z(x, y, z) = tau_z(x, -y, z)`. Averaging correctly transforms both components.

**PR #482 (thorfinn) already proved that applying mirror-y augmentation during training hurts** (EP5 gate fail, val 10.989%). That version polluted the loss with both augmented and original geometries simultaneously. This PR is different: **zero training changes** — the model trains identically to SOTA, and TTA is applied only at eval/test time.

This is a pure inference trick: it costs exactly 2× forward passes at eval time (which adds ~20 min to a 270-min run, so ~7% overhead), but does not add any training cost, parameters, or risk of divergence. Near-zero downside.

## Instructions

All changes go in `target/trainer_runtime.py`. The training loop is NOT touched — the only changes are to `accumulate_eval_batch` (or a new wrapper around it) and the call sites where `evaluate_split` is invoked.

### Step 1: Add CLI flag

In `target/train.py`, add a CLI flag:

```python
tta_mirror_y: bool = False  # --tta-mirror-y flag
```

Pass it through to `evaluate_split` via config.

### Step 2: Create a mirror-y batch transform helper

Add a helper function in `trainer_runtime.py` (near `accumulate_eval_batch`):

```python
def mirror_y_batch(batch: SurfaceBatch) -> SurfaceBatch:
    """
    Return a new SurfaceBatch with y-axis mirrored geometry.
    
    Coordinate layout (from data/loader.py):
      surface_x: [xyz(0:3), normals(3:6), area(6)]
      surface_y: [surface_pressure(0), tau_x(1), tau_y(2), tau_z(3)]
      volume_x:  [xyz(0:3), sdf(3)]
      volume_y:  [volume_pressure(0)]
    
    Mirror-y transform rules:
      - Flip y coordinate:  surface_x[:, 1] → -surface_x[:, 1]
      - Flip normal_y:      surface_x[:, 4] → -surface_x[:, 4]
      - Negate tau_y:       surface_y[:, 2] → -surface_y[:, 2]
        (tau_y is antisymmetric under mirror-y: τ_y(x,-y,z) = -τ_y(x,y,z))
      - tau_x, tau_z, surface_pressure, volume_pressure: unchanged
        (symmetric under mirror-y)
      - Volume: flip y coordinate and normal_y if present
        volume_x[:, 1] → -volume_x[:, 1]  (y coord of volume points)
    """
    # Deep clone tensors (don't modify originals)
    sx = batch.surface_x.clone()
    sy = batch.surface_y.clone()
    vx = batch.volume_x.clone()
    vy = batch.volume_y.clone()
    
    # Mirror surface x: flip y coord and normal_y
    sx[..., 1] = -sx[..., 1]   # y → -y
    sx[..., 4] = -sx[..., 4]   # n_y → -n_y
    
    # Mirror surface y: tau_y flips sign
    sy[..., 2] = -sy[..., 2]   # tau_y → -tau_y
    
    # Mirror volume x: flip y coord only (no normals in volume_x)
    vx[..., 1] = -vx[..., 1]   # y → -y
    
    return SurfaceBatch(
        surface_x=sx,
        surface_y=sy,
        surface_mask=batch.surface_mask.clone(),
        volume_x=vx,
        volume_y=vy,
        volume_mask=batch.volume_mask.clone(),
        case_ids=batch.case_ids,
    )
```

### Step 3: Implement TTA in accumulate_eval_batch

Modify `accumulate_eval_batch` to optionally average predictions over original + mirrored:

```python
def accumulate_eval_batch(
    accumulator: EvalAccumulator,
    *,
    model: nn.Module,
    batch: SurfaceBatch,
    transform: TargetTransform,
    device: torch.device,
    amp_mode: str,
    tta_mirror_y: bool = False,
) -> None:
    batch = batch.to(device)
    surface_target_norm = transform.apply_surface(batch.surface_y)
    volume_target_norm = transform.apply_volume(batch.volume_y)
    eval_module = unwrap_model(model)
    with autocast_context(device, amp_mode):
        out = eval_module(
            surface_x=batch.surface_x,
            surface_mask=batch.surface_mask,
            volume_x=batch.volume_x,
            volume_mask=batch.volume_mask,
        )
    
    if tta_mirror_y:
        # Run the mirrored batch and average predictions
        mirrored = mirror_y_batch(batch)
        with autocast_context(device, amp_mode):
            out_mirror = eval_module(
                surface_x=mirrored.surface_x,
                surface_mask=mirrored.surface_mask,
                volume_x=mirrored.volume_x,
                volume_mask=mirrored.volume_mask,
            )
        # Mirror-back the mirrored predictions:
        # surface tau_y needs to be re-negated (it was predicted in mirrored space)
        # surface_pressure, tau_x, tau_z, volume_pressure: symmetric, just average
        sp_mirror = out_mirror["surface_preds"].float()
        sp_mirror[..., 2] = -sp_mirror[..., 2]  # un-negate tau_y
        
        surface_pred_norm = (out["surface_preds"].float() + sp_mirror) / 2.0
        volume_pred_norm = (out["volume_preds"].float() + out_mirror["volume_preds"].float()) / 2.0
    else:
        surface_pred_norm = out["surface_preds"].float()
        volume_pred_norm = out["volume_preds"].float()
    
    # Rest of accumulate_eval_batch unchanged from here...
```

### Step 4: Thread tta_mirror_y through evaluate_split

Update `evaluate_split` to accept and pass through the `tta_mirror_y` flag:

```python
def evaluate_split(
    model: nn.Module,
    loader,
    transform: TargetTransform,
    device: torch.device,
    *,
    amp_mode: str = "none",
    distributed_state: DistributedState | None = None,
    tta_mirror_y: bool = False,
) -> dict[str, float]:
    model.eval()
    accumulator = EvalAccumulator()
    for batch in loader:
        accumulate_eval_batch(
            accumulator,
            model=model,
            batch=batch,
            transform=transform,
            device=device,
            amp_mode=amp_mode,
            tta_mirror_y=tta_mirror_y,
        )
    # ... rest unchanged
```

Pass `tta_mirror_y=config.tta_mirror_y` wherever `evaluate_split` is called from the training loop (look for the `evaluate_split(` call sites in `train.py`).

### Step 5: Run on SOTA stack

Train the full SOTA config (unchanged — training is NOT modified). After training, the best-val checkpoint is evaluated with TTA enabled. If you want to see the delta directly without retraining, also run a post-hoc TTA eval:

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm --rff-num-features 16 \
  --tta-mirror-y \
  --wandb-group fern-tta-mirror-y \
  --wandb-name fern/inference-time-mirror-y-tta
```

### Step 6: Sanity check

At EP1-2, val_abupt should be in the same ballpark as the SOTA stack (TTA doesn't change training). If you see a large divergence from EP1, the mirror_y_batch transform may have a sign error.

Also check: does tau_y (val) improve vs no-TTA? If yes, the symmetry averaging is working. If tau_y regresses, there may be a sign error in the un-negation step.

## Baseline

Current SOTA: **PR #387 alphonse** (feat16 RFF + QK-norm + STRING-sep)
- **val_abupt: 7.3816%** (EP11, W&B run `wj6mn6ve`)
- **test_abupt: 8.5936%**

Per-axis test metrics (must beat):
| Metric | SOTA #387 | AB-UPT |
|---|---:|---:|
| surface_pressure | 4.4377% | 3.82% |
| wall_shear | 7.9989% | 7.29% |
| volume_pressure | 12.1885% | 6.08% |
| tau_x | 6.9622% | 5.35% |
| **tau_y** | **9.1058%** | **3.65%** ← primary target |
| **tau_z** | **10.2736%** | **3.63%** ← secondary target |

**Merge bar: val_abupt < 7.3816%.**

**Note:** PR #482 (thorfinn) proved mirror-y augmentation DURING training hurts (+3.6pp at EP5). This PR is explicitly inference-only and does not touch the training loop.

### EP5 gate

Report val_abupt EP5 ≤ 9.2%. TTA overhead adds ~20 min to each eval pass so the run may be EP11-capped rather than EP12. Report the best-val checkpoint metrics as usual.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Checking in — I see PR #499 is open for the inference-time mirror-y TTA experiment but I cannot find any W&B runs launched under this experiment. 

Please implement and launch the mirror-y TTA experiment as described in the PR. To recap the approach: at inference time, flip each car geometry across the y-axis (mirror image), run it through the model, then average the predictions with the original orientation predictions. This exploits the approximate bilateral symmetry of the DrivAerML geometries and should reduce prediction variance on asymmetric flow features.

Steps to implement:
1. In the inference/evaluation code, add a flag `--tta_mirror_y` 
2. When enabled: for each sample, create a y-mirrored copy of the point cloud (multiply all y-coordinates by -1), run through model, average with original predictions
3. Evaluate on the full validation set with and without TTA and report both val_abupt numbers

Please launch the run and post the W&B link in this PR. The W&B run name should include `tay-fern-tta-mirror` and use `--wandb_group fern-tta-mirror` for grouping.

**Comment by morganmcg1** [comment]
ADVISOR: Fern, I still don't see any W&B runs or commits on this branch. Please prioritize getting this experiment moving — TTA is a low-risk, high-leverage variance reduction that we expect to help on the asymmetric tau_y/tau_z channels especially.

Heads up on a known issue you'll hit: PR #488's multi-sigma STRING-sep init code was never actually committed (only BASELINE.md was updated). The current head doesn't have the `--rff-init-sigmas` flag. Two options:

1. **Recommended for TTA**: Skip the multi-sigma flag entirely for this experiment. TTA is orthogonal to the positional encoding init — just train on the current head with the existing reproduce command (drop `--rff-init-sigmas`) and add the `--tta_mirror_y` evaluation path on top. This isolates the TTA effect cleanly.
2. If you want the multi-sigma init too, askeladd is patching it inline in PR #516 — wait for that to land or coordinate.

Concrete next steps:
1. `git fetch origin && git checkout fern-tta-mirror-y` (or rebase onto tay)
2. Add `--tta_mirror_y` flag to `train.py` Config + parse_args
3. In `trainer_runtime.py` `accumulate_eval_batch`/`evaluate_split`, when `tta_mirror_y` is set: run a second forward on a y-mirrored batch (negate y-coord, n_y, tau_y), un-mirror predictions (negate predicted tau_y), average with the original-orientation prediction, then compute val metrics on the averaged prediction.
4. Train with the standard reproduce command minus `--rff-init-sigmas`, evaluate twice — once with TTA off, once with TTA on — and report both val_abupt numbers and per-axis (especially tau_y/tau_z).
5. `--wandb_group fern-tta-mirror`

Please post a status update + W&B link within the next pod cycle.

**Comment by morganmcg1** [comment]
STUDENT fern: Status update — TTA experiment is in flight.

**Implementation:**
- `mirror_y_batch(batch)` helper in `trainer_runtime.py`: flips `surface_x[..., 1]` (y), `surface_x[..., 4]` (n_y), `surface_y[..., 2]` (tau_y), and `volume_x[..., 1]` (volume y); leaves x, z, n_x, n_z, area, sdf, p_s, tau_x, tau_z, p_v unchanged.
- `accumulate_eval_batch(..., tta_mirror_y=False)`: when enabled, runs a second forward on the mirrored batch, un-negates `surface_preds[..., 2]` (tau_y) on the mirrored prediction, then averages with the original prediction. Volume preds are averaged directly (volume_pressure is symmetric).
- `evaluate_split(..., tta_mirror_y=False)` threads the flag through.
- `train.py`: `--tta-mirror-y` CLI flag added; passed to all `evaluate_split` call sites.
- `run_final_evaluation`: when TTA is on, does a second `full_val`/`test` pass with TTA off and logs under `full_val_no_tta_*` / `test_no_tta_*` namespaces (only takes effect on relaunches).

**Symmetry math (sanity check):** Under mirror about xz-plane: $u_y$ flips sign, $u_x, u_z$ unchanged → $\\sigma_{xy}, \\sigma_{yz}$ are antisymmetric, $\\sigma_{xx}, \\sigma_{yy}, \\sigma_{zz}, \\sigma_{xz}$ symmetric → wall traction $T_y$ flips sign, $T_x, T_z$ unchanged. So tau_y is the only antisymmetric channel, matching the implementation.

**Run status (W&B group `fern-tta-mirror-y`, run name `fern/inference-time-mirror-y-tta`):**
W&B run IDs (DDP rank 0–7): `dy3viqmk`, `v4f44z5z`, `3hwj4n3y`, `s9igpbfo`, `r4sy7cl0`, `305pyvv4`, `smnil313`, `o5dlndfp`. Rank 0 = `dy3viqmk`.

Started 06:04 UTC. Trajectory so far (val_abupt with TTA on):
| Epoch | val_abupt |
|---:|---:|
| 1 | 51.34 |
| 2 | 30.05 |
| 3 | 14.47 |
| 4 | 11.18 |
| 5 | 10.02 |
| 6 |  9.54 |

EP5 gate (≤9.2%) not met but trajectory is converging toward expected SOTA-stack territory. Training ETA: 10:34 UTC timeout, then full val + test eval until ~12:04 UTC.

**Plan for both TTA-on/off numbers:** Since the running process loaded `trainer_runtime.py` before I added the TTA-on/off double pass, the in-run final eval will only log TTA-on metrics. After training completes, I'll run `post_hoc_eval.py` to load the best checkpoint and evaluate both ways on the same model, then resume the rank-0 W&B run (`dy3viqmk`) and log under `posthoc_tta_on_*` / `posthoc_tta_off_*`.

Will report final numbers when the run finishes.

**Comment by morganmcg1** [comment]
ADVISOR: Implementation looks correct — the symmetry math is right (tau_y is the only antisymmetric surface channel under xz-plane mirror). Good sanity check.

EP5 gate miss (10.02% > 9.2%) is not a concern at this stage — trajectory is converging normally (EP6=9.54%). The gate was set for early-exit safety, not as a quality ceiling. Continue to completion.

**Post-hoc eval plan confirmed:** After the timeout at ~10:34 UTC, evaluate best-val checkpoint twice with `post_hoc_eval.py` and log under `posthoc_tta_on_*` / `posthoc_tta_off_*` to rank-0 run `dy3viqmk`. Report:
1. TTA-on vs TTA-off val_abupt + full per-axis
2. Delta on tau_y specifically (this is the primary TTA signal we're watching for)
3. Test metrics from best-val checkpoint (TTA-on)

The key question: does TTA-on reduce val_tau_y or val_tau_z compared to TTA-off on the same model? If yes, even by 0.2-0.3pp, that's a portable inference-time gain that can stack on top of any future training improvement.

Continue.

**Comment by morganmcg1** [comment]
ADVISOR: Fern — your posthoc TTA comparison data has been retrieved from W&B run `dy3viqmk` (best-val checkpoint). The result is **conclusively negative for inference-time mirror-y TTA on this stack**.

## Comparison table (best-val checkpoint, full_val set, 34 cases)

| Metric | TTA ON (training-eval) | TTA OFF (posthoc) | Δ (TTA hurts) |
|---|---:|---:|---:|
| **val_abupt** | **8.834%** | **7.657%** | **+1.177pp** |
| surface_pressure | 6.308% | 4.984% | +1.324pp |
| wall_shear | 9.583% | 8.643% | +0.940pp |
| wall_shear_x | 8.289% | 7.506% | +0.783pp |
| **wall_shear_y (tau_y)** | **10.804%** | **9.693%** | **+1.110pp** |
| **wall_shear_z (tau_z)** | **12.819%** | **11.512%** | **+1.307pp** |
| volume_pressure | 5.953% | 4.591% | +1.362pp |

## Interpretation

The hypothesis was that mirror-y TTA would reduce variance on the antisymmetric tau_y channel. **The data refutes this on every axis** — including the channels we expected to gain (tau_y, tau_z) and the channels expected to be neutral (sp, vp). TTA hurts *uniformly*.

Most likely explanation: the training loop ran validation with TTA on, so the best-val checkpoint was selected against TTA-blended predictions. When TTA is removed at inference, the unaided model is actually stronger. This means TTA was acting as label smoothing during training-validation, masking better single-pass predictions.

## Posthoc TTA-OFF result vs current SOTA

| | val_abupt | test_abupt |
|---|---:|---:|
| Fern TTA OFF (posthoc) | 7.657% | 8.613% |
| Current SOTA #489 (r5rw40rn) | 7.179% | 8.497% |
| Δ | +0.478pp | +0.116pp |

Doesn't beat SOTA on either metric, so TTA itself yields no win, but the OFF result is much closer to SOTA than the ON one.

## Decision: closing this PR as a clear negative result

- TTA mirror-y at inference does not help on any axis with this training protocol
- We will not pursue inference-time mirror-y TTA further
- Lesson: any future TTA experiment must use TTA-OFF validation for checkpoint selection (so the saved weights are optimal for unaided inference). Or train with TTA-on as a *training-time* augmentation rather than inference-time

## Next assignment

Fresh hypothesis incoming via separate assignment PR. Closing this one.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #499 because Inference-time mirror-y TTA hurts on every axis (val_abupt: TTA ON 8.834% vs TTA OFF 7.657%, +1.177pp). Posthoc data conclusively negative. Documented in PR comments for future reference. Closing as dead end — no follow-up TTA experiments planned. Fresh assignment will follow..
